### PR TITLE
Prepare for concurrent IOVs, L1ConfigOnlineProdBase

### PR DESCRIPTION
**Pull request by wddgit**
Prepare for concurrent IOVs in files related to L1ConfigOnlineProdBase
and L1ConfigOnlineProdBaseExt. Mostly this is conversion of
shared_ptr to unique_ptr in those base classes and the many ESProducers
that derive from them. In a few modules I needed to remove a data member
to do this.

Note this also required changing the return type of the deserialization
function in CondCore/CondDB/interface/Serialization.h and the fetchPayload
function in CondCore/CondDB/interface/Session.h. Both of these changes
seem to be better design to me. Probably the only reason it was a shared_ptr
was because this code was originally written before unique_ptr existed. I do
not know if it is significant, but this should also improve performance at least
slightly. (One could make more similar changes in CondCore, but I only did
the minimum necessary for my immediate goal).

There is also a bit of cleanup mixed in. "scram code-checks" fixed a few things.
In one of the previous PRs in this ESProducer campaign
someone requested I remove all the unnecessary "using edm::es" namespace
directives, so I removed all those in these packages as well. Plus a little bit of
fixing indentation around the lines I changed.

FYI @Dr15Jones

### CODE REVIEW

**Comment by wddgit** [comment]
please test

**Comment by wddgit** [comment]
please test

**Comment by wddgit** [comment]
please test

rebased to remove conflicts

File: L1TriggerConfig/GMTConfigProducers/src/L1MuGMTParametersProducer.cc [hunk 1]
```diff
@@ -34,8 +34,6 @@ L1MuGMTParametersProducer::~L1MuGMTParametersProducer() {
 std::unique_ptr<L1MuGMTParameters> 
 L1MuGMTParametersProducer::produceL1MuGMTParameters(const L1MuGMTParametersRcd& iRecord)
 {
-  using namespace edm::es;
-
   std::unique_ptr<L1MuGMTParameters> gmtparams = std::unique_ptr<L1MuGMTParameters>( new L1MuGMTParameters() );
```
Comment (thomreis) [inline L1TriggerConfig/GMTConfigProducers/src/L1MuGMTParametersProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/GMTConfigProducers/src/L1MuGMTParametersProducer.cc [hunk 2]
```diff
@@ -87,8 +85,6 @@ L1MuGMTParametersProducer::produceL1MuGMTParameters(const L1MuGMTParametersRcd&
 std::unique_ptr<L1MuGMTChannelMask> 
 L1MuGMTParametersProducer::produceL1MuGMTChannelMask(const L1MuGMTChannelMaskRcd& iRecord)
 {
-  using namespace edm::es;
-
   std::unique_ptr<L1MuGMTChannelMask> gmtchanmask = std::unique_ptr<L1MuGMTChannelMask>( new L1MuGMTChannelMask() );
```
Comment (thomreis) [inline L1TriggerConfig/GMTConfigProducers/src/L1MuGMTParametersProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/L1ScalesProducers/src/L1MuGMTScalesProducer.cc [hunk 3]
```diff
@@ -66,8 +66,6 @@ L1MuGMTScalesProducer::~L1MuGMTScalesProducer() {}
 std::unique_ptr<L1MuGMTScales> 
 L1MuGMTScalesProducer::produceL1MuGMTScales(const L1MuGMTScalesRcd& iRecord)
 {
-   using namespace edm::es;
-
    std::unique_ptr<L1MuGMTScales> l1muscale = std::unique_ptr<L1MuGMTScales>( new L1MuGMTScales( m_scales ) );
```
Comment (thomreis) [inline L1TriggerConfig/L1ScalesProducers/src/L1MuGMTScalesProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/L1ScalesProducers/src/L1MuTriggerPtScaleProducer.cc [hunk 4]
```diff
@@ -31,8 +31,6 @@ L1MuTriggerPtScaleProducer::~L1MuTriggerPtScaleProducer() {}
 std::unique_ptr<L1MuTriggerPtScale> 
 L1MuTriggerPtScaleProducer::produceL1MuTriggerPtScale(const L1MuTriggerPtScaleRcd& iRecord)
 {
-   using namespace edm::es;
-
    std::unique_ptr<L1MuTriggerPtScale> l1muscale =
      std::unique_ptr<L1MuTriggerPtScale>( new L1MuTriggerPtScale( m_scales ) );
```
Comment (thomreis) [inline L1TriggerConfig/L1ScalesProducers/src/L1MuTriggerPtScaleProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/L1ScalesProducers/src/L1MuTriggerScalesProducer.cc [hunk 5]
```diff
@@ -58,8 +58,6 @@ L1MuTriggerScalesProducer::~L1MuTriggerScalesProducer() {}
 std::unique_ptr<L1MuTriggerScales> 
 L1MuTriggerScalesProducer::produceL1MuTriggerScales(const L1MuTriggerScalesRcd& iRecord)
 {
-   using namespace edm::es;
-
    std::unique_ptr<L1MuTriggerScales> l1muscale =
      std::unique_ptr<L1MuTriggerScales>( new L1MuTriggerScales( m_scales ) );
```
Comment (thomreis) [inline L1TriggerConfig/L1ScalesProducers/src/L1MuTriggerScalesProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc [hunk 6]
```diff
@@ -53,17 +53,13 @@ L1ScalesTrivialProducer::~L1ScalesTrivialProducer()
 // ------------ method called to produce the data  ------------
 std::unique_ptr<L1CaloEtScale> L1ScalesTrivialProducer::produceEmScale(const L1EmEtScaleRcd& iRecord)
 {
-   using namespace edm::es;
-
    std::unique_ptr<L1CaloEtScale> emScale = std::unique_ptr<L1CaloEtScale>( new L1CaloEtScale(m_emEtScaleInputLsb, m_emEtThresholds) );
```
Comment (thomreis) [inline L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc [hunk 7]
```diff
@@ -53,17 +53,13 @@ L1ScalesTrivialProducer::~L1ScalesTrivialProducer()
...
 
    return emScale ;
 }
 
 std::unique_ptr<L1CaloEtScale> L1ScalesTrivialProducer::produceJetScale(const L1JetEtScaleRcd& iRecord)
 {
-   using namespace edm::es;
-
    std::unique_ptr<L1CaloEtScale> jetEtScale = std::unique_ptr<L1CaloEtScale>( new L1CaloEtScale(m_jetEtScaleInputLsb, m_jetEtThresholds) );
```
Comment (thomreis) [inline L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc [hunk 8]
```diff
@@ -72,8 +68,6 @@ std::unique_ptr<L1CaloEtScale> L1ScalesTrivialProducer::produceJetScale(const L1
 
 std::unique_ptr<L1CaloEtScale> L1ScalesTrivialProducer::produceHtMissScale(const L1HtMissScaleRcd& iRecord)
 {
-   using namespace edm::es;
-
    std::unique_ptr<L1CaloEtScale> htMissScale = std::unique_ptr<L1CaloEtScale>( new L1CaloEtScale(0, 0x7f, m_jetEtScaleInputLsb, m_htMissThresholds) );
```
Comment (thomreis) [inline L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc]: Maybe use `std::make_unique` instead?

File: L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc [hunk 9]
```diff
@@ -82,8 +76,6 @@ std::unique_ptr<L1CaloEtScale> L1ScalesTrivialProducer::produceHtMissScale(const
 
 std::unique_ptr<L1CaloEtScale> L1ScalesTrivialProducer::produceHfRingScale(const L1HfRingEtScaleRcd& iRecord)
 {
-   using namespace edm::es;
-
    std::unique_ptr<L1CaloEtScale> hfRingEtScale = std::unique_ptr<L1CaloEtScale>( new L1CaloEtScale(0xff, 0x7, m_jetEtScaleInputLsb, m_hfRingThresholds) );
```
Comment (thomreis) [inline L1TriggerConfig/L1ScalesProducers/src/L1ScalesTrivialProducer.cc]: Maybe use `std::make_unique` instead?

**Comment by wddgit** [comment]
please test

@thomreis I made all the changes you suggested and pushed them in the latest commit. Thanks, that improves the code.

**Comment by fabiocos** [comment]
@ggovi @lpernie @tocheng could you please check this PR? It looks mostly a technical update

**Comment by ggovi** [comment]
+1

File: CondCore/CSCPlugins/src/plugin.cc [hunk 10]
```diff
@@ -62,17 +62,17 @@
...
 
 namespace cond {
-  template <> std::shared_ptr<CSCReadoutMapping> deserialize<CSCReadoutMapping>( const std::string& payloadType,
-										   const Binary& payloadData,
-										   const Binary& streamerInfoData ){
+  template <> std::unique_ptr<CSCReadoutMapping> deserialize<CSCReadoutMapping>( const std::string& payloadType,
+                                                                                 const Binary& payloadData,
+                                                                                 const Binary& streamerInfoData ){
     // DESERIALIZE_BASE_CASE( CSCReadoutMapping ); abstract
     DESERIALIZE_POLIMORPHIC_CASE( CSCReadoutMapping, CSCReadoutMappingFromFile );
```
Comment (ggovi) [inline CondCore/CSCPlugins/src/plugin.cc]: the macro DESERIALIZE_POLIMORPHIC_CASE returns a std::shared_ptr<T> .  Are you here converting it in std::unique_ptr<T>?  Is this thing compiling at all?  Or am I missing something?

Comment (wddgit) [reply CondCore/CSCPlugins/src/plugin.cc]: Yes, the return types are changed from shared_ptr to unique_ptr for the deserialize function and everything related to that which is either internal to deserialize or in the call stack up to ESProducer produce functions. When the DESERIALIZE_POLIMORPHIC_CASE is used there is an implicit conversion of the pointer type (and a move to another unique_ptr). Everything in the CMSSW repository compiles (see PR test build and the PR tests all passed). The only reason to have a design where deserialize returns a shared_ptr would be if it or something internal to it actually shares ownership and management of the memory of the allocated object. It currently does not and I cannot imagine ever modifying deserialize to do that in the future. As far as code that uses the deserialize function, it is easy to convert a unique_ptr to a shared_pt if you ever want to do that at any higher level. But it is problematic to ever change from a shared_ptr to a unique_ptr at a higher level. These pointers get passed up all the way to the output of ESProducer produce functions. If the produce functions in ESProducers return shared_ptr's, then there will be problems when we start processing multiple IOVs concurrently. That is the motivation for this PR. I hope that explains it. Let me know if there are more questions or discussion. (I did highlight this in the comments at the top of the PR as I understand this is an interface change in something important and low level).

Comment (ggovi) [reply CondCore/CSCPlugins/src/plugin.cc]: I confirm that these changes would not compile without the corresponding change in CondCore/CondDB/Serialization.h, or a re-impelmentation of the macro.  I can be wrong, but I don't find the file in your list.  
Anyhow, discussing in the context of the validation of a single PR with the title "Prepare for concurrent IOVs" a drastic change in the CondDB interface, like the move from shared_ptr to unique_ptr does not sounds to me appropriate. 
I'm afraid you will have to revert this change.

**Comment by wddgit** [comment]
I think you are right. This PR should be split. I can make the change in Serialization.h and the other low level code in a separate PR with a different title. That is a very reasonable request. Better than what I did. I will do that on Monday. I would suggest we postpone our discussion and a decision whether to approve it or not until I submit it. That discussion probably belongs with that PR.

The other part of this PR will just have to wait for that new PR to be approved before I resubmit it. As you say, very little compiles without the low level changes. For the simple reason that one cannot convert a shared_ptr to a unique_ptr (at least not in a nice way).

I am not sure I would call this a drastic change though. I cannot see a reason for the return type of deserialize to be shared_ptr in the first place. The code internal to deserialize is not managing the memory. It does not use the extra abilities of shared_ptr. So this change just removes something unused. Also it is very easy to convert a unique_ptr to a shared_ptr. Direct assignment works. Very little needs to be changed in higher level code. Most of the higher level code will work without modification if the lower level code is changed. (We may decide we also want to change the higher level code, but that is a different story).

Beyond the issues associated with running multiple IOVs concurrently, there are also performance issues that would motivate us to make this change. shared_ptr requires a reference counter or maybe two. If make_shared isn't used they usually are allocated separately. Also increments and decrements to the counters must be synchronized across threads.  I do not know if it is significant here, but shared _ptr's are expensive compared to unique_ptr. The performance of a unique_ptr is almost as good as a raw pointer and it is also simpler to use and think about.

Maybe I am missing something, but I see all good and no downsides to this.

FYI @Dr15Jones

**Comment by Dr15Jones** [comment]
The use of `std::unique_ptr` makes it trivial for us to know that a given ESProducer or ESSource will properly work with concurrent IOVs. That is because the data product object can not be held by the ESProducer/ESSource since the `std::unique_ptr` forces single object ownership, which in this case is the EventSetup system itself. If a `std::shared_ptr` is used, it is not clear if the ESProducer/ESSource maintains a hold on the same object and then will attempt to reuse that object on a second IOV. If it did so, then since the object is still being used by the first IOV we would have a problem if the ESProducer/ESSource were to change the object. 

Therefore switching from an `std::shared_ptr` to an `std::unique_ptr` aids future maintenance for when we switch to concurrent IOVs since we can be absolutely certain the object is not accidently shared between different IOVs.

**Pull request closed without merging**